Add detailed guide for building ros2_rust packages

nnmm opened the pull request:
Also extend contributing.md and condense the readme a bit. See [here](https://github.com/ros2-rust/ros2_rust/blob/doc_articles/README.md) for a rendered version.

jhdcs reviewed (changes requested):
Only a partial review at the moment, I'll have to look at the rest later...

esteve commented on README.md:
````diff
@@ -11,94 +8,50 @@ Build status
+```shell
+mkdir -p workspace/src && cd workspace
 git clone https://github.com/ros2-rust/ros2_rust.git src/ros2_rust
+docker build -t ros2_rust_dev - < src/ros2_rust/Dockerfile
````
I don't agree with recommending using Docker for developing `ros2_rust`. I use it myself, but the preferred way should be more in line with the rest of ROS2, which does not require Docker at all. I'm ok with adding a section about how to use Docker with `ros2_rust`, but should be in addition to explaining a Docker-less workflow.

nnmm replied:
Sure. Do you mean that the method in the README should be the non-Docker way, or only that I should explain the non-Docker way in `Building.md`? I assumed that most people looking at the README would be interested in a quick and non-intrusive way of trying it out, hence I used the Docker setup.

esteve replied:
I think recommending Docker as the way to develop `ros2-rust` packages or the library itself is more intrusive, as it requires setting up a piece of software that is not strictly needed for ROS2 development. If someone has set up a ROS2 workspace already, they already have everything they need, and don't require Docker.

So, what I suggest is that we keep the Docker-less workflow, and then add an optional chapter about how to use Docker with `ros2-rust` for those of us who prefer to use it.

nnmm replied:
@esteve Done.

nnmm commented on docs/Building.md:
````diff
@@ -0,0 +1,214 @@
+git clone https://github.com/ros2-rust/ros2_rust.git src/ros2_rust
+```
+
+*Note: Once `rclrs` is published on crates.io, it's not technically needed anymore to clone the `ros2_rust` repo, and this section should be adapted to reflect that.*
````
Done.

nnmm commented on README.md:
````diff
@@ -11,94 +8,50 @@ Build status
 ```
 
-Enjoy!
+For an actual guide, see the following documents:
````
This section has changed a bit, but I've kept the "more in-depth guide" wording.

nnmm commented on README.md:
```diff
@@ -11,94 +8,50 @@ Build status
-
-- The `rclrs` interface is very limited for now and might not be idiomatic yet, any help and suggestion on the interface would be greatly appreciated
-- Due to the current ROS2 support of non-default clients, packages containing definitions of messages used in Rust crates must be present in the current workspace; otherwise message crates generation won't be triggered
+The client library is still rapidly evolving, and there are no stability guarantees.
```
Done.

nnmm commented on README.md:
```diff
@@ -11,94 +8,50 @@ Build status
-pip install git+https://github.com/colcon/colcon-cargo.git git+https://github.com/colcon/colcon-ros-cargo.git
-# Install the cargo-ament-build plugin
-cargo install cargo-ament-build
+In a nutshell, the steps to get started are:
```
This should be solved now, the latest version of README.md doesn't require Docker.

jhdcs commented on docs/Building.md:
```diff
@@ -0,0 +1,230 @@
+## Environment setup
+
+Building `rclrs` requires a standard [ROS 2 installation](https://docs.ros.org/en/foxy/Installation.html), and a few extensions.
+These extensions are: `colcon-cargo`, `colcon-ros-cargo`, `cargo-ament-build`. The former two are `colcon` plugins, and the latter is a `cargo` plugin.
```
```suggestion
These extensions are: `colcon-cargo`, `colcon-ros-cargo`, `cargo-ament-build`. The first two are `colcon` plugins, and the third is a `cargo` plugin.
```
English is odd. Nothing's technically wrong with how you have it, but I think this looks a bit more "natural". Feel free to ignore, though.

jhdcs commented on docs/Building.md:
```diff
@@ -0,0 +1,230 @@
+
+### Choosing a workspace directory and cloning `ros2_rust`
+
+A "workspace directory", or just "workspace", is simply a directory of your choosing that contains your `ros2_rust` checkout and potentially other ROS 2 packages. It will also usually be your working directory for building. There is only one limitation: It must not contain the ROS 2 installation, so it can't be `/`, for instance. Note that this has nothing to do with a [`cargo` workspace](https:/ ...
```
```suggestion
A "workspace directory", or just "workspace", is simply a directory of your choosing that contains your `ros2_rust` checkout and potentially other ROS 2 packages. It will also usually be your working directory for building. There is only one limitation: It must not contain the ROS 2 installation, so it can't be `/`, for instance. Note that this has **nothing** to do with a [`cargo` workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).
```
A little extra emphasis to attract eyes to this sentence. Hopefully that will make it more likely that users/readers won't miss this and get confused.

jhdcs commented on docs/Building.md:
```diff
@@ -0,0 +1,230 @@
+
+*Note: Once `rclrs` is published on crates.io, it's not technically needed anymore to clone the `ros2_rust` repo, and this section will be modified to reflect that.*
+
+### Using the Docker image
```
```suggestion
### ALTERNATIVE: Using the Docker image
```
This suggestion may not be needed.

nnmm replied:
You're right that it wasn't clear that this is not always required. I restructured it so that the "manual installation" and "docker" sections are next to each other, and marked as alternatives.

jhdcs commented on docs/Building.md:
```diff
@@ -0,0 +1,230 @@
+
+### Package identification
+
+When you forgot to source the ROS 2 installation, you'll get this error:
```
```suggestion
If you forget to source the ROS 2 installation, you'll get this error:
```

nnmm replied:
Thanks, good catch!

jhdcs reviewed (changes requested):
Just a few more suggested changes. That being said, I like how this is turning out!

jhdcs commented on docs/Building.md:
```diff
@@ -0,0 +1,233 @@
+
+### Failed to resolve patches
+
+When you've deleted your `install` directory but not your `.cargo` directory, you'll get this error:
```
```suggestion
If you've deleted your `install` directory but not your `.cargo` directory, you'll get this error:
```

nnmm replied:
Done!

jhdcs reviewed (changes requested):
I apologize for not catching this one earlier...

nnmm commented:
Thanks @jhdcs! Any more changes requested from your side, @esteve?

esteve commented on docs/Contributing.md:
```diff
@@ -1,4 +1,4 @@
+# Contributing to `ros2_rust`
 Contributions to `ros2_rust` are very welcome!
 
 Work is coordinated in the [Matrix chat](https://matrix.to/#/+rosorg-rust:matrix.org). It's a good idea to check in with the other contributors there before starting development on a new feature.
```
While Matrix is nice, I personally prefer if the focus is on GitHub. It's hard to keep track of conversations on Matrix, and there's no trail of who is working on what. So I'd advise contributors to first check GitHub, file an issue for what they want to work on, and then perhaps talk about it on Matrix, but again, technical discussions on a chat are hard to follow.

nnmm replied:
Ok, I rephrased this.

nnmm commented on docs/Building.md:
```diff
@@ -0,0 +1,230 @@
+sudo apt install -y git libclang-dev python3-pip python3-vcstool # libclang-dev is required by bindgen
+# Install these plugins for cargo and colcon:
+cargo install cargo-ament-build
+pip install git+https://github.com/colcon/colcon-cargo.git git+https://github.com/colcon/colcon-ros-cargo.git
```
Sure, done

esteve commented:
@nnmm the `Contributing.md` file should go back to the root and renamed back to `CONTRIBUTING.md`. GitHub uses that convention so that contributors get a notification before they submit code (see https://github.blog/2012-09-17-contributing-guidelines/). This is in accordance with other projects not specific to ROS2 as well (e.g. https://github.com/django/django/blob/main/CONTRIBUTING.rst, https://github.com/pocoproject/poco/blob/master/CONTRIBUTING.md)

nnmm commented:
> the Contributing.md file should go back to the root and renamed back to CONTRIBUTING.md. GitHub uses that convention so that contributors get a notification before they submit code (see https://github.blog/2012-09-17-contributing-guidelines/). This is in accordance with other projects not specific to ROS2 as well (e.g. https://github.com/django/django/blob/main/CONTRIBUTING.rst, https://github.com/pocoproject/poco/blob/master/CONTRIBUTING.md)

@esteve if you look up a newer article, GitHub also checks the `docs` subdirectory. I'd also be surprised if it needs to be uppercase, but I renamed it just to make sure.

esteve commented:
@nnmm you're right about the `docs/` folder (https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors), but in that article seems to always refer to the file as `CONTRIBUTING(.md)`, so perhaps it needs to be in uppercase anyway.

nnmm commented:
Yep, it's uppercase now.

esteve commented:
@nnmm that was all, thanks for addressing my feedback :slightly_smiling_face:
